### PR TITLE
Homework 04. Tram finite state machine

### PR DESCRIPTION
**Pull request by dmitry-sidorov**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,4 +11,29 @@
...
+  Stop_by_railway_block ---> |continue the route| Move_to_station
+  Move_to_station ---> |traffic light red| Stop_by_traffic_light
+  Stop_by_traffic_light ---> |traffic light green| Move_to_station
+  Move_to_station ---> |unexpected accident happened| Stop_by_accident
+  Stop_by_accident ---> |open the doors| Emergency_doors_opening
+  Emergency_doors_opening --->|start rescue passengers| Rescue_passengers
+  Rescue_passengers ---> |stop rescue passengers| Emergency_doors_opening
+  Emergency_doors_opening ---> |close the doors| Stop_by_accident
+  Stop_by_accident ---> |move to service station| Service
+  Service ---> |back to depot| Depot
```
Comment (am-kantox) [inline README.md]: OMG

You don’t like gradual approaches, do you? :) A bit of general advice, while I didn’t get further to read the implementation: start with implementing “main” states, which are 5 tops, then explore each of them. Einstein supposedly said (well, this is usually attributed to him,) that the normal human can operate 5 different entities at a time, and for those geniuses, this number is 7. Don’t prove the humanity is stupid by inventing the code requiring readers to manipulate 10+ :)

---

Also, I’d say you sometimes confuse _tram_ states and their internal attributes. What about keeping _whys_ in attributes and _whats_ in the state itself? Like (I am oversimplifying it, but still)
```elixir
in_depot → |move| moving
moving → |stop| on_stop # conditional
moving → |bang| on_accident
on_stop → |doors| on_stop # conditional
on_stop → |move| moving # this checks if doors are closes etc
on_stop → |move| to_depot # final state
on_accident → |doors| on_accident
on_accident → |repair| moving
```
Notice `move` is _the very same event_ accepted in two states (`in_depot` and `on_stop`) _and_ resulting in two states (`moving`, `to_depot`.)

File: tram_fsm/lib/tram_fsm.ex [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+  def info(pid) do
+    GenServer.call(pid, :info)
+  end
+
+  def transition_scheme do
+    @transition_scheme
+  end
+
+  defp get_available_transitions(%{tram_state: current_state}) do
```
Comment (am-kantox) [inline tram_fsm/lib/tram_fsm.ex:54]: ❤❤❤

File: tram_fsm/lib/tram_fsm.ex [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    @transition_scheme
+    |> Map.to_list()
+    |> Enum.filter(fn {_transition, %{from: from}} -> from == current_state end)
```
Comment (am-kantox) [inline tram_fsm/lib/tram_fsm.ex]:
```suggestion
    |> Enum.filter(&match?({_, %{from: ^current_state}, &1}))
```

Comment (am-kantox) [inline tram_fsm/lib/tram_fsm.ex]: Or
```suggestion
    |> Enum.filter(fn 
      {_transition, %{from: ^current_state}} -> true
      _ -> false
     end)
```

File: tram_fsm/lib/tram_fsm.ex [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    |> Enum.map(fn {transition, _} -> transition end)
+  end
+
+  defp get_info(state) do
+    %{
+      current_state: state[:tram_state],
```
Comment (am-kantox) [inline tram_fsm/lib/tram_fsm.ex]:
```suggestion
      current_state: state.tram_state,
```
There _is_ a difference between a dot notation, `Map.get/3`, and `Access` (which is effectively applied on `[key]` call. The first two are nearly the same, `Access` is way more generic and hence slower. Whenever possible, use dot/explicit get.

File: tram_fsm/lib/tram_fsm.ex [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      available_transitions: get_available_transitions(state)
+    }
+  end
+
+  def transition(pid, transition) do
+    case Map.fetch(@transition_scheme, transition) do
```
Comment (am-kantox) [inline tram_fsm/lib/tram_fsm.ex]: ❤

File: tram_fsm/lib/tram_fsm.ex [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  end
+
+  @impl true
+  def handle_call(:info, _, state) do
+    {:reply, {:info, get_info(state)}, state}
+  end
+
+  @impl true
+  def handle_call({:transition, %{from: from, to: to}}, _, state)
+      when from === state.tram_state do
```
Comment (am-kantox) [inline tram_fsm/lib/tram_fsm.ex]:
```suggestion
  def handle_call({:transition, %{from: from, to: to}}, _, %{tram_state: from} = state) do
```

Also, `===/2` does not make sense for anything but numbers, use `==/2`, it’s ok :)

File: tram_fsm/lib/tram_fsm.ex [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    new_state = %{tram_state: to}
+    {:reply, {:ok, new_state}, new_state}
+  end
+
+  @impl true
+  def handle_call({:transition, %{to: to}}, _, state)
+      when to === state.tram_state do
```
Comment (am-kantox) [inline tram_fsm/lib/tram_fsm.ex]:
```suggestion
  def handle_call({:transition, %{to: to}}, _, %{tram_state: to} = state) do
```

**Review by am-kantox — APPROVED** [review]
:zap: :droplet: :zap: :gem: :zap: :blue_heart: :zap:

Great job. A couple of cosmetic notes I left should not make you think it could have been done better, cosmetic stuff can always be improved.

I have the only doubt that the initial state diagram might have been done simpler, but   it’s a matter of taste as well.

File: tram_fsm/lib/tram_fsm.ex (hunk 6 above)
Comment (dmitry-sidorov) [reply tram_fsm/lib/tram_fsm.ex]: I have JavaScript background, my bad :) kind of muscle memory

File: README.md (hunk 1 above)
Comment (dmitry-sidorov) [reply README.md]: Yep, 7±2 numbers are also called "Miller's wallet". Valid point. I'll try to reduce reading complexity of my code in the next iteration. :)

I thought, transition name should be unique for some reason 😄 So there is at least one way to make my code simpler. 🎉

File: tram_fsm/lib/tram_fsm.ex (hunk 3 above)
Comment (dmitry-sidorov) [reply tram_fsm/lib/tram_fsm.ex]: I like this one!

File: tram_fsm/lib/tram_fsm.ex (hunk 7 above)
Comment (dmitry-sidorov) [reply tram_fsm/lib/tram_fsm.ex]: no `==` ~~operator~~ function -> no problem with `===` 😆 

```
_def handle_call({:transition, %{to: to}}, _, %{tram_state: to} = state) do_
```
if right-side `to`'s are not the same value, will pattern matching return an error? I mean the case when:
```
%{to: "moving"}
%{tram_state: "stop"}
```

Comment (am-kantox) [reply tram_fsm/lib/tram_fsm.ex]: > will pattern matching return an error

It won’t match, yes.

File: README.md [hunk 8]
```diff
@@ -1,39 +1,39 @@
...
-  Stop_by_accident ---> |open the doors| Emergency_doors_opening
-  Emergency_doors_opening --->|start rescue passengers| Rescue_passengers
-  Rescue_passengers ---> |stop rescue passengers| Emergency_doors_opening
-  Emergency_doors_opening ---> |close the doors| Stop_by_accident
-  Stop_by_accident ---> |move to service station| Service
-  Service ---> |back to depot| Depot
-```
\ No newline at end of file
+  In_depot ---> |move| Moving
+  Moving ---> |move| In_depot
```
Comment (am-kantox) [inline README.md]: One should not decide what the next state is from outside. What if the current state is `Moving` _and_ one sends `move` event? The tram all of a sudden finds itself in the depot, which is not exactly how trams and depots function (unless this is a tram that knows how to null-transport itself to the depot.) 

Instead, the tram itself should figure out whether it’s the depot all around, or a regular stop, upon receiving `stop` event. 

```suggestion
  Moving ---> |stop| In_depot
```

File: README.md [hunk 9]
```diff
@@ -1,39 +1,39 @@
...
+  Moving ---> |stop| On_station
+  On_station ---> |move| Moving
+  On_station ---> |doors| Onboarding_passengers
+  Onboarding_passengers ---> |doors| On_station
```
Comment (am-kantox) [inline README.md]: I’d say one of these states is redundant.
```suggestion
  Moving ---> |stop| Onboarding
  Onboarding ---> |move| Moving
```

File: README.md [hunk 10]
```diff
@@ -1,39 +1,39 @@
...
+  Moving ---> |block| Blocked
+  Blocked ---> |block| Moving
+  Moving ---> |traffic light| On_red_light_stop
```
Comment (am-kantox) [inline README.md]: Well, I’d also call the event `stop`.

**Review by am-kantox — COMMENTED** [review]
Уже отлично, даже по сути придраться не к чему :)

File: README.md (hunk 10 above)
Comment (dmitry-sidorov) [reply README.md]: fixed

File: README.md (hunk 8 above)
Comment (dmitry-sidorov) [reply README.md]: done

File: README.md (hunk 9 above)
Comment (dmitry-sidorov) [reply README.md]: fixed

**Pull request merged**